src/main/java/net/minecraftforge/fml/common/registry/VillagerRegistry.java

```diff
@@ -306,10 +392,54 @@ public boolean equals(Object o)
      */
     public static void setRandomProfession(EntityVillager entity, Random rand)
     {
-        Set<String> entries = INSTANCE.professions.getKeys();
+        ArrayList<String> entries = new ArrayList(INSTANCE.professionsByName.keySet());
         int prof = rand.nextInt(entries.size());
-        //TODO: Grab id range from internal registry
-        entity.setProfession(rand.nextInt(5));
+        VillagerProfession profession = INSTANCE.professionsByName.get(entries.get(prof));
+        entity.setProfession(profession.id);
```

Comment:
Problem with this is that HashMaps are not ordered and since this is used in world gen things will most likely derp and you dont keep vanilla world gen compatibility, you need a ordered list of professions by ID.